src/base/bittorrent/torrentdescriptor.cpp

```diff
@@ -0,0 +1,232 @@
+BitTorrent::TorrentDescriptor::loadFromFile(const Path &path) noexcept
+try
+{
+    return TorrentDescriptor(lt::load_torrent_file(path.toString().toStdString(), loadTorrentLimits()));
+}
+catch (const lt::system_error &err)
+{
+    return nonstd::make_unexpected(QString::fromLocal8Bit(err.what()));
+}
+
+nonstd::expected<BitTorrent::TorrentDescriptor, QString>
+BitTorrent::TorrentDescriptor::parse(const QString &str) noexcept
+try
+{
+    QString magnetURI = str;
+    if (isV2Hash(str))
+        magnetURI = u"magnet:?xt=urn:btmh:1220" + str; // 0x12 0x20 is the "multihash format" tag for the SHA-256 hashing scheme.
+    else if (isV1Hash(str))
+        magnetURI = u"magnet:?xt=urn:btih:" + str;
+
+    return TorrentDescriptor(lt::parse_magnet_uri(magnetURI.toStdString()));
+}
+catch (const lt::system_error &err)
+{
+    return nonstd::make_unexpected(QString::fromLocal8Bit(err.what()));
+}
+
+nonstd::expected<void, QString> BitTorrent::TorrentDescriptor::saveToFile(const Path &path) const
+try
+{
+    const lt::entry torrentEntry = lt::write_torrent_file(m_ltAddTorrentParams);
+    const nonstd::expected<void, QString> result = Utils::IO::saveToFile(path, torrentEntry);
+    if (!result)
+        return result.get_unexpected();
+
+    return {};
+}
+catch (const lt::system_error &err)
+{
+    return nonstd::make_unexpected(QString::fromLocal8Bit(err.what()));
+}
+
+BitTorrent::TorrentDescriptor::TorrentDescriptor(lt::add_torrent_params ltAddTorrentParams)
+    : m_ltAddTorrentParams {std::move(ltAddTorrentParams)}
+{
+    if (m_ltAddTorrentParams.ti && m_ltAddTorrentParams.ti->is_valid())
+        m_info.emplace(*m_ltAddTorrentParams.ti);
+}
+
+BitTorrent::InfoHash BitTorrent::TorrentDescriptor::infoHash() const
+{
+#ifdef QBT_USES_LIBTORRENT2
+    return m_ltAddTorrentParams.info_hashes;
+#else
+    return m_ltAddTorrentParams.info_hash;
+#endif
+}
+
+QString BitTorrent::TorrentDescriptor::name() const
+{
+    return m_info ? m_info->name() : QString::fromStdString(m_ltAddTorrentParams.name);
+}
+
+QDateTime BitTorrent::TorrentDescriptor::creationDate() const
+{
+    return ((m_ltAddTorrentParams.ti->creation_date() != 0)
+            ? QDateTime::fromSecsSinceEpoch(m_ltAddTorrentParams.ti->creation_date()) : QDateTime());
+}
+
+QString BitTorrent::TorrentDescriptor::creator() const
+{
+    return QString::fromStdString(m_ltAddTorrentParams.ti->creator());
+}
+
+QString BitTorrent::TorrentDescriptor::comment() const
+{
+    return QString::fromStdString(m_ltAddTorrentParams.ti->comment());
+}
+
+const std::optional<BitTorrent::TorrentInfo> &BitTorrent::TorrentDescriptor::info() const
+{
+    return m_info;
+}
+
+void BitTorrent::TorrentDescriptor::setTorrentInfo(TorrentInfo torrentInfo)
+{
+    if (!torrentInfo.isValid())
+    {
+        m_info.reset();
+        m_ltAddTorrentParams.ti.reset();
+    }
+    else
+    {
+        m_info = std::move(torrentInfo);
+        m_ltAddTorrentParams.ti = torrentInfo.nativeInfo();
```

Comment:
@glassez 
coverity scan detected a 'use after move' here.

Reply:
Fixed by #19397.
Thank you!